crates/swc_ecma_lints/src/rules/no_use_before_define.rs

```diff
@@ -0,0 +1,233 @@
+        };
+
+        if let Some(spans) = self.scoped_spans.get_mut(&spans_block_id) {
+            spans.insert(span);
+        } else {
+            let mut spans: AHashSet<Span> = Default::default();
+
+            spans.insert(span);
+
+            self.scoped_spans.insert(spans_block_id, spans);
+        }
+    }
+
+    fn has_access_before_define(&self, id: &Id) -> bool {
+        let current_scope = self.scope.last().unwrap();
+
+        self.scoped_indents.get(current_scope).unwrap().contains(id)
+    }
+
+    fn get_ident_spans_in_current_scope(&self, id: &Id) -> &AHashSet<Span> {
+        let current_block = self.scope.last().unwrap();
+
+        let spans_block_id = SpansScopeId {
+            ident_id: id.clone(),
+            block_span: *current_block,
+        };
+
+        self.scoped_spans.get(&spans_block_id).unwrap()
+    }
+
+    fn emit_report(&self, es6_var_check: bool, span: Span, name: &str) {
+        let message = format!("'{}' was used before it was defined", name);
+
+        let expected_reaction = if es6_var_check {
+            LintRuleReaction::Error
+        } else {
+            self.expected_reaction
+        };
+
+        HANDLER.with(|handler| match expected_reaction {
+            LintRuleReaction::Error => {
+                handler.struct_span_err(span, &message).emit();
+            }
+            LintRuleReaction::Warning => {
+                handler.struct_span_warn(span, &message).emit();
+            }
+            _ => {}
+        });
+    }
+
+    fn check_ident(&self, es6_var_check: bool, ident: &Ident) {
+        let ident_id = ident.to_id();
+
+        if self.has_access_before_define(&ident_id) {
+            let sym = &*ident.sym;
+            let mut spans = self
+                .get_ident_spans_in_current_scope(&ident_id)
+                .iter()
+                .copied()
+                .collect::<Vec<Span>>();
+
+            spans.sort();
+
+            spans.into_iter().for_each(|span| {
+                self.emit_report(es6_var_check, span, sym);
+            });
+        }
+    }
+
+    fn check_pat(&self, es6_var_check: bool, pat: &Pat) {
+        match pat {
+            Pat::Ident(BindingIdent { id, .. }) => {
+                self.check_ident(es6_var_check, id);
+            }
+            Pat::Array(ArrayPat { elems, .. }) => {
+                elems.iter().for_each(|elem| {
+                    if let Some(elem) = elem {
+                        self.check_pat(es6_var_check, elem);
+                    }
+                });
+            }
+            Pat::Object(ObjectPat { props, .. }) => {
+                props.iter().for_each(|prop| match prop {
+                    ObjectPatProp::Assign(AssignPatProp { key, .. }) => {
+                        self.check_ident(es6_var_check, key);
+                    }
+                    ObjectPatProp::KeyValue(KeyValuePatProp { value, .. }) => {
+                        self.check_pat(es6_var_check, value.as_ref());
```

Comment:
I think we need `self.check_ident(es6_var_check, key);` here, as `key` of `KeyValuePatProp` creates a binding, just like `Pat::Ident`

Reply:
```
function f() {
  const {
    x: { y }
  } = { x: { y: 10, z: 20 } }

  console.log(x); // ReferenceError: x is not defined
}

f();
```